src/Microsoft.IdentityModel.Protocols.SignedHttpRequest/SignedHttpRequestHandler.cs

```diff
@@ -0,0 +1,1417 @@
+        {
+            if (signedHttpRequestValidationContext == null)
+                throw LogHelper.LogArgumentNullException(nameof(signedHttpRequestValidationContext));
+
+            if (signedHttpRequestValidationContext.SignedHttpRequestValidationParameters.PopKeysResolverAsync != null)
+                return await signedHttpRequestValidationContext.SignedHttpRequestValidationParameters.PopKeysResolverAsync(signedHttpRequest, validatedAccessToken, signedHttpRequestValidationContext, cancellationToken).ConfigureAwait(false);
+
+            var cnf = GetCnfClaimValue(signedHttpRequest, validatedAccessToken, signedHttpRequestValidationContext);
+            var popKey = await ResolvePopKeyFromCnfClaimAsync(cnf, signedHttpRequest, validatedAccessToken, signedHttpRequestValidationContext, cancellationToken).ConfigureAwait(false);
+            return new List<SecurityKey>() { popKey };
+        }
+
+        /// <summary>
+        /// Gets the JSON representation of the 'cnf' claim.
+        /// This method expects a "cnf" claim to be present as a claim of the <paramref name="validatedAccessToken"/> ("at").
+        /// </summary>
+        /// <param name="signedHttpRequest">A signed http request as a JWT.</param>
+        /// <param name="validatedAccessToken">An access token ("at") that was already validated during the SignedHttpRequest validation process.</param>
+        /// <param name="signedHttpRequestValidationContext">A structure that wraps parameters needed for SignedHttpRequest validation.</param>
+        /// <returns>JSON representation of the 'cnf' claim.</returns>
+        protected virtual string GetCnfClaimValue(SecurityToken signedHttpRequest, SecurityToken validatedAccessToken, SignedHttpRequestValidationContext signedHttpRequestValidationContext)
+        {
+            if (validatedAccessToken == null)
+                throw LogHelper.LogArgumentNullException(nameof(validatedAccessToken));
+
+            if (!(validatedAccessToken is JsonWebToken jwtValidatedAccessToken))
+                throw LogHelper.LogExceptionMessage(new SignedHttpRequestValidationException(LogHelper.FormatInvariant(LogMessages.IDX23031, validatedAccessToken.GetType(), typeof(JsonWebToken), validatedAccessToken)));
+
+            // use the decrypted jwt if the jwtValidatedAccessToken is encrypted.
+            if (jwtValidatedAccessToken.InnerToken != null)
+                jwtValidatedAccessToken = jwtValidatedAccessToken.InnerToken;
+            
+            if (jwtValidatedAccessToken.TryGetPayloadValue(ConfirmationClaimTypes.Cnf, out JObject cnf) && cnf != null)
+                return cnf.ToString();
+            else if (jwtValidatedAccessToken.TryGetPayloadValue(ConfirmationClaimTypes.Cnf, out string cnfString) && !string.IsNullOrEmpty(cnfString))
+                return cnfString;
```

Comment:
Perf consideration:
GetCnfClaimValue returns string. Either directly from the cnf claim or from the JObject.ToString.

Then, this string is again parsed to JObject in ResolvePopKeyFromCnfClaimAsync.

What about returning the JObject from GetCnfClaimValue directly?

Reply:
That's right, great catch.
These methods were initially protected, and we didn't want to expose Newtonsoft objects.
I'll make the change.

Reply:
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1260/commits/4ad150d3bb53a80d672b8afda8202e70eeaa13a8